File: backend/src/AgentTeam/Services/MemoryAutoUpdater.php
<?php

declare(strict_types=1);

namespace AgentTeam\Services;

use PDO;

/**
 * Orchestrates the post-turn memory auto-update:
 *   skip cheap cases → extract → merge per scope → compact if over budget → write + event row.
 *
 * Designed to be called after fastcgi_finish_request() so the user never
 * waits on the extractor call. Any failure is swallowed and logged — the
 * assistant's response has already been delivered.
 */
class MemoryAutoUpdater
{
    public const MIN_USER_MSG_CHARS = 20;

    private PDO $db;
    private string $apiKey;
    private UserMemoryRepository $memoryRepo;
    private UserMemoryEventsRepository $eventsRepo;
    private UserMemorySettingsRepository $settingsRepo;

    public function __construct(PDO $db, string $apiKey)
    {
        $this->db = $db;
        $this->apiKey = $apiKey;
        $this->memoryRepo = new UserMemoryRepository($db);
        $this->eventsRepo = new UserMemoryEventsRepository($db);
        $this->settingsRepo = new UserMemorySettingsRepository($db);
    }

    /**
     * Run the auto-update flow for one chat turn. Returns an array describing
     * what happened (for tests/debug) or null if short-circuited.
     */
    public function run(int $userId, string $sessionId, string $lastUserMsg, string $lastAssistantMsg): ?array
    {
        if ($userId <= 0 || $this->apiKey === '') {
            return null;
        }
        if (mb_strlen(trim($lastUserMsg)) < self::MIN_USER_MSG_CHARS) {
            return null;
        }

        $settings = $this->settingsRepo->get($userId);
        if (!$settings['enabled']) {
            return null;
        }

        $current = $this->memoryRepo->getBoth($userId);
        $currentMemory = $current[UserMemoryRepository::SCOPE_MEMORY];
        $currentUser = $current[UserMemoryRepository::SCOPE_USER];

        $extractor = new MemoryExtractor($this->apiKey);
        $result = $extractor->extract(
            $settings['model'],
            $currentMemory,
            $currentUser,
            $lastUserMsg,
            $lastAssistantMsg
        );

        if ($result === null) {
            return null;
        }

        $written = [];

        // Second-stage guard: filter paraphrased/implied duplicates against current content.
        // Only runs when there are candidates, so no overhead on the common empty case.
        $memoryAdditions = !empty($result['memory_additions'])
            ? $extractor->filterDuplicates($settings['model'], $currentMemory, $result['memory_additions'])
            : [];
        $userAdditions = !empty($result['user_additions'])
            ? $extractor->filterDuplicates($settings['model'], $currentUser, $result['user_additions'])
            : [];

        if (!empty($memoryAdditions)) {
            $this->applyScope(
                $userId,
                UserMemoryRepository::SCOPE_MEMORY,
                $currentMemory,
                $memoryAdditions,
                $result['reason'],
                $sessionId,
                $settings['model'],
                $extractor,
                $written
            );
        }

        if (!empty($userAdditions)) {
            $this->applyScope(
                $userId,
                UserMemoryRepository::SCOPE_USER,
                $currentUser,
                $userAdditions,
                $result['reason'],
                $sessionId,
                $settings['model'],
                $extractor,
                $written
            );
        }

        return [
            'reason' => $result['reason'],
            'written' => $written,
        ];
    }

    private function applyScope(
        int $userId,
        string $scope,
        string $current,
        array $additions,
        string $reason,
        string $sessionId,
        string $model,
        MemoryExtractor $extractor,
        array &$written
    ): void {
        $budget = $this->memoryRepo->budgetFor($scope);
        $candidate = $this->memoryRepo->buildMerged($current, $additions);

        if ($candidate === $current) {
            return; // all additions dedup'd away
        }

        if (mb_strlen($candidate) <= $budget) {
            $this->memoryRepo->set($userId, $scope, $candidate);
            $this->eventsRepo->log(
                $userId,
                $scope,
                UserMemoryEventsRepository::SOURCE_AUTO_EXTRACT,
                $current,
                $candidate,
                $reason,
                $sessionId
            );
            $written[] = $scope;
            return;
        }

        // Over budget — ask the model to compact, preserving the new additions verbatim.
        $compacted = $extractor->compact($model, $candidate, $budget, $additions);
        if ($compacted === null || $compacted === '') {
            error_log("[MemoryAutoUpdater] compact returned empty for user={$userId} scope={$scope}; skipping write");
            return;
        }

        $this->memoryRepo->set($userId, $scope, $compacted);
        $this->eventsRepo->log(
            $userId,
            $scope,
            UserMemoryEventsRepository::SOURCE_COMPACT,
            $current,
            $compacted,
            $reason,
            $sessionId
        );
        $written[] = $scope;
    }
}

File: backend_python/app/agent_team/services/memory_auto_updater.py
"""Port of backend/src/AgentTeam/Services/MemoryAutoUpdater.php.

Orchestrates the post-turn memory auto-update:
  skip cheap cases → extract → merge per scope → compact if over budget → write + event row.

Designed to be called after the response is already delivered to the user
so the extractor call never adds latency. Any failure is swallowed and
logged — the assistant's response has already been delivered.
"""
from __future__ import annotations

from app.agent_team.services.memory_extractor import MemoryExtractor
from app.agent_team.services.user_memory_events_repository import UserMemoryEventsRepository
from app.agent_team.services.user_memory_repository import UserMemoryRepository
from app.agent_team.services.user_memory_settings_repository import UserMemorySettingsRepository
from app.support.logger import error_log
from app.support.phpcompat import php_empty


class MemoryAutoUpdater:
    MIN_USER_MSG_CHARS = 20

    def __init__(self, db, apiKey: str):
        self.db = db
        self.apiKey = apiKey
        self.memoryRepo = UserMemoryRepository(db)
        self.eventsRepo = UserMemoryEventsRepository(db)
        self.settingsRepo = UserMemorySettingsRepository(db)

    def run(self, userId: int, sessionId: str, lastUserMsg: str, lastAssistantMsg: str) -> dict | None:
        """Run the auto-update flow for one chat turn. Returns a dict describing
        what happened (for tests/debug) or None if short-circuited."""
        if userId <= 0 or self.apiKey == '':
            return None
        if len(lastUserMsg.strip()) < self.MIN_USER_MSG_CHARS:
            return None

        settings = self.settingsRepo.get(userId)
        if not settings['enabled']:
            return None

        current = self.memoryRepo.getBoth(userId)
        currentMemory = current[UserMemoryRepository.SCOPE_MEMORY]
        currentUser = current[UserMemoryRepository.SCOPE_USER]

        extractor = MemoryExtractor(self.apiKey)
        try:
            result = extractor.extract(
                settings['model'],
                currentMemory,
                currentUser,
                lastUserMsg,
                lastAssistantMsg,
            )

            if result is None:
                return None

            written: list = []

            # Second-stage guard: filter paraphrased/implied duplicates against current content.
            # Only runs when there are candidates, so no overhead on the common empty case.
            memoryAdditions = (
                extractor.filterDuplicates(settings['model'], currentMemory, result['memory_additions'])
                if not php_empty(result['memory_additions'])
                else []
            )
            userAdditions = (
                extractor.filterDuplicates(settings['model'], currentUser, result['user_additions'])
                if not php_empty(result['user_additions'])
                else []
            )

            if not php_empty(memoryAdditions):
                self._applyScope(
                    userId,
                    UserMemoryRepository.SCOPE_MEMORY,
                    currentMemory,
                    memoryAdditions,
                    result['reason'],
                    sessionId,
                    settings['model'],
                    extractor,
                    written,
                )

            if not php_empty(userAdditions):
                self._applyScope(
                    userId,
                    UserMemoryRepository.SCOPE_USER,
                    currentUser,
                    userAdditions,
                    result['reason'],
                    sessionId,
                    settings['model'],
                    extractor,
                    written,
                )

            return {
                'reason': result['reason'],
                'written': written,
            }
        finally:
            # Python-only addition: PHP has no equivalent — Guzzle clients
            # die with the request. Release the extractor's httpx client
            # regardless of how run() exits (early return, exception).
            extractor.close()

    def _applyScope(
        self,
        userId: int,
        scope: str,
        current: str,
        additions: list,
        reason: str,
        sessionId: str,
        model: str,
        extractor: MemoryExtractor,
        written: list,
    ) -> None:
        budget = self.memoryRepo.budgetFor(scope)
        candidate = self.memoryRepo.buildMerged(current, additions)

        if candidate == current:
            return  # all additions dedup'd away

        if len(candidate) <= budget:
            self.memoryRepo.set(userId, scope, candidate)
            self.eventsRepo.log(
                userId,
                scope,
                UserMemoryEventsRepository.SOURCE_AUTO_EXTRACT,
                current,
                candidate,
                reason,
                sessionId,
            )
            written.append(scope)
            return

        # Over budget — ask the model to compact, preserving the new additions verbatim.
        compacted = extractor.compact(model, candidate, budget, additions)
        if compacted is None or compacted == '':
            error_log(f'[MemoryAutoUpdater] compact returned empty for user={userId} scope={scope}; skipping write')
            return

        self.memoryRepo.set(userId, scope, compacted)
        self.eventsRepo.log(
            userId,
            scope,
            UserMemoryEventsRepository.SOURCE_COMPACT,
            current,
            compacted,
            reason,
            sessionId,
        )
        written.append(scope)
